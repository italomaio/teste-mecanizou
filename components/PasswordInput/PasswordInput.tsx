"use client";

import React, { forwardRef, useMemo } from "react";
import { Button, Input } from "@/components";
import { Eye, EyeClosed, LucideProps } from "lucide-react";

type PasswordInputProps = React.ComponentProps<"input">;

const iconProps: LucideProps = {
  className: "text-sapphire dark:text-sapphire-50",
  size: 18,
};

const PasswordInput = forwardRef<HTMLInputElement, PasswordInputProps>(
  ({ className, ...props }, ref) => {
    const [showPassword, setShowPassword] = React.useState<boolean>(false);

    const togglePassword = () => {
      setShowPassword((prev) => !prev);
    };

    const Icon = useMemo(() => {
      const Component = !showPassword ? Eye : EyeClosed;
      return <Component {...iconProps} />;
    }, [showPassword]);

    return (
      <div className="relative flex items-center">
        <Input
          ref={ref}
          type={showPassword ? "text" : "password"}
          className={className}
          {...props}
        />
        <Button
          type="button"
          size="icon"
          variant="ghost"
          className="absolute right-1"
          onClick={togglePassword}
        >
          {Icon}
        </Button>
      </div>
    );
  }
);

PasswordInput.displayName = "PasswordInput";

export default PasswordInput;
