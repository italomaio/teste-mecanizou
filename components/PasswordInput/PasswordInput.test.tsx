import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import PasswordInput from "./PasswordInput";

describe("PasswordInput", () => {
  it("Should render the input field", () => {
    render(<PasswordInput />);
    const input = screen.getByRole("textbox");
    expect(input).toBeInTheDocument();
  });

  it("Should render the password toggle button", () => {
    render(<PasswordInput />);
    const button = screen.getByRole("button");
    expect(button).toBeInTheDocument();
  });

  it("Should hide the password by default", () => {
    render(<PasswordInput />);
    const input = screen.getByRole("textbox");
    expect(input).toHaveAttribute("type", "password");
  });

  it("Should toggle the password visibility when the button is clicked", () => {
    render(<PasswordInput />);
    const input = screen.getByRole("textbox");
    const button = screen.getByRole("button");

    // Initially hidden
    expect(input).toHaveAttribute("type", "password");

    // Click to show password
    fireEvent.click(button);
    expect(input).toHaveAttribute("type", "text");

    // Click to hide password again
    fireEvent.click(button);
    expect(input).toHaveAttribute("type", "password");
  });

  it("Should pass additional props to the input field", () => {
    render(<PasswordInput placeholder="Enter your password" />);
    const input = screen.getByPlaceholderText("Enter your password");
    expect(input).toBeInTheDocument();
  });

  it("Should apply the provided className to the input field", () => {
    render(<PasswordInput className="custom-class" />);
    const input = screen.getByRole("textbox");
    expect(input).toHaveClass("custom-class");
  });

  it("Should match the snapshot", () => {
    const { asFragment } = render(<PasswordInput />);
    expect(asFragment()).toMatchSnapshot();
  });
});
