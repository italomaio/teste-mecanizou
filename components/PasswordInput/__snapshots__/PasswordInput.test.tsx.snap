// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PasswordInput Should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="relative flex items-center"
  >
    <input
      class="flex h-9 w-full rounded-sm bg-transparent px-3 py-1 text-base shadow-sm transition-colors placeholder:text-muted-foreground disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-sapphire text-gray-700 border border-sapphire file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground"
      role="textbox"
      type="password"
    />
    <button
      class="rounded focus:outline-none inline-flex items-center justify-center gap-2 transition-colors duration-300 cursor-pointer font-medium text-md hover:bg-accent hover:text-accent-foreground p-2 text-base absolute right-1"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-eye text-sapphire dark:text-sapphire-50"
        fill="none"
        height="18"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M2.062 12.348a1 1 0 0 1 0-.696 10.75 10.75 0 0 1 19.876 0 1 1 0 0 1 0 .696 10.75 10.75 0 0 1-19.876 0"
        />
        <circle
          cx="12"
          cy="12"
          r="3"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
